### ISSUE TITLE
Release polars objects?

### ISSUE THREAD

**Issue by eitsupi** [rpolars/r-polarssql#2]
I am wondering if there is a way to safely release polars objects such as DataFrame.
I am currently overriding it to reference another DataFrame and not actually releasing the resource.

https://github.com/rpolars/r-polarssql/blob/c985255bb82e8e2e2e834b00011458a46f24bd78/R/dbClearResult_Result.R#L9-L11

(`res@env$resultset` is polars DataFrame here)

@sorhawell Could you take a look at this?

**Comment by sorhawell** [comment]
Yes I will write some examples tonight

**Comment by sorhawell** [comment]
``` r
  library(polars) # built on branch release_me

# Levels of abstraction between RPolarsDataFrame and arrow srray(s)
#1 An R variable can point to R SEXP external ptr. (variable references + garbage collection!)
#2 SEXP ExtPtr can point to (and own) an RPolarsDataFrame
#3 a RPolarsDataFrame is thin wrapper around rust polars DataFrame
#4 rust polars DataFrame is struct containing Vec<Series>
#5 Series is as Arc<dyn SeriesTrait>.  (Arc is reference counting!)
#6 dyn Series trait is runtime polymorphism over any ChunkedArray<T> where T is the DataType
#7 ChunkedArray<T> struct has field called "chunks "which is Vec<ArrayRef>
#8 ArrayRef is trait that that interact with some Arrow array.
#9 Arrow array "producer" can give strong references to "consumers" managed by (reference counting!)


# example one, release via R gc() - level 1
df = polars::pl$DataFrame(mtcars)
df = NULL # either overwrite variable
rm(df) # ... or remove it completely
gc()
#>           used (Mb) gc trigger (Mb) limit (Mb) max used (Mb)
#> Ncells  705863 37.7    1404404 75.1         NA   976191 52.2
#> Vcells 1349656 10.3    8388608 64.0      16384  2270467 17.4
# gc() will find no variables pointing to ExtPtr and will destoy it. The will cascade release 1-9, 
# if no other strong references existed through out.
# sorry we cannot impl Drop for DataFrame to debug print when released because of some rust rules.


# example two, release directly via r-polars - level 3-5
df2 = polars::pl$DataFrame(mtcars)
df2_copy = df2 # make a new variable pointing to the same ExtPtr
df2_clone = df2$clone() # make a new unique DataFrame which has it own strong references to Series(s)
s2 = df2$get_column("cyl") # clone a Series within the DataFrame, increment reference on "cyl" Series


# drop all columns in place, this is a bit quirky
#lapply(df2$columns, .pr$DataFrame$drop_in_place, self = df2)[c()] #drop each column in turn level 4-5

# new method from branch release_me, just drops pl::DataFrame inside RPolarsDataFrame and replaces 
# with an empty. **NOTE 1
.pr$DataFrame$drop_all_in_place(df2) # we make this method and instead drop in place - level 3-5
print(df2) # all columns are removed from df
#> shape: (0, 0)
#> ┌┐
#> ╞╡
#> └┘
print(df2_copy) # also from here
#> shape: (0, 0)
#> ┌┐
#> ╞╡
#> └┘
print(df2_clone) # not from here
#> shape: (32, 11)
#> ┌──────┬─────┬───────┬───────┬───┬─────┬─────┬──────┬──────┐
#> │ mpg  ┆ cyl ┆ disp  ┆ hp    ┆ … ┆ vs  ┆ am  ┆ gear ┆ carb │
#> │ ---  ┆ --- ┆ ---   ┆ ---   ┆   ┆ --- ┆ --- ┆ ---  ┆ ---  │
#> │ f64  ┆ f64 ┆ f64   ┆ f64   ┆   ┆ f64 ┆ f64 ┆ f64  ┆ f64  │
#> ╞══════╪═════╪═══════╪═══════╪═══╪═════╪═════╪══════╪══════╡
#> │ 21.0 ┆ 6.0 ┆ 160.0 ┆ 110.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
#> │ 21.0 ┆ 6.0 ┆ 160.0 ┆ 110.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
#> │ 22.8 ┆ 4.0 ┆ 108.0 ┆ 93.0  ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0  ┆ 1.0  │
#> │ 21.4 ┆ 6.0 ┆ 258.0 ┆ 110.0 ┆ … ┆ 1.0 ┆ 0.0 ┆ 3.0  ┆ 1.0  │
#> │ …    ┆ …   ┆ …     ┆ …     ┆ … ┆ …   ┆ …   ┆ …    ┆ …    │
#> │ 15.8 ┆ 8.0 ┆ 351.0 ┆ 264.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 4.0  │
#> │ 19.7 ┆ 6.0 ┆ 145.0 ┆ 175.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 6.0  │
#> │ 15.0 ┆ 8.0 ┆ 301.0 ┆ 335.0 ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 8.0  │
#> │ 21.4 ┆ 4.0 ┆ 121.0 ┆ 109.0 ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0  ┆ 2.0  │
#> └──────┴─────┴───────┴───────┴───┴─────┴─────┴──────┴──────┘
s2 # neither from here
#> polars Series: shape: (32,)
#> Series: 'cyl' [f64]
#> [
#>  6.0
#>  6.0
#>  4.0
#>  6.0
#>  8.0
#>  6.0
#>  8.0
#>  4.0
#>  4.0
#>  6.0
#>  6.0
#>  8.0
#>  …
#>  4.0
#>  4.0
#>  8.0
#>  8.0
#>  8.0
#>  8.0
#>  4.0
#>  4.0
#>  4.0
#>  8.0
#>  6.0
#>  8.0
#>  4.0
#> ]

#**NOTE1 one may ask why is example TWO any different from example ONE. The difference is that
# it does NOT require to call gc(). The flipside, mutable objects in R unwanted complexity,
# and should be hidden from the public API.

#conclusion it is possible to release arrow arrays if one hold the only strong reference on level 
#1, 5, & 9.
# either during gc() or by explicitly dropping the DataFrame / Series on a rust level or even
# lower levels.

# potentially references can also be hold to individual chunks on the arrow level, then those chunks
# alone will be protected from release

# It is not possible to force release overriding strong references in a safe way.

# Reccomendation
# If the inner DataFrame was not shared with anyone, there are many ways to release. If someone
# else holds strong references, only a partial release is possible at best.

#1 consider that the field can either be NULL or a RPolarsDataFrame. Free simply setting the field to NULL and wait for gc()

#or adopt

# compatible with today polars
drop_df_in_place = function(df) {
  lapply(df$columns, polars::.pr$DataFrame$drop_in_place, self = df)
  NULL
}
df = pl$DataFrame(mtcars)
drop_df_in_place(df)
#> NULL


# tiny bit faster after release_me is merged, but same same
drop_df_in_place = function(df) {
  polars::.pr$DataFrame$drop_all_in_place(df)
}
df = pl$DataFrame(mtcars)
drop_df_in_place(df)
```

<sup>Created on 2023-11-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by eitsupi** [comment]
Thanks for the detailed description! Very helpful as I understand very little about the internals of Polars.

Perhaps the ability to release DataFrames would be useful.

**Comment by eitsupi** [comment]
Closed by 0cd1fb3d56d4d76a50bfcbad464a96a42db1e73d
